### DISCUSSION TITLE
This Maui Controls will be the Most Popular Someday

### DISCUSSION THREAD

**Discussion by nextcodelab** [dotnet/Microsoft.Maui.Graphics.Controls#47] [General]
I just discovered it now,, I really dont know that this library exist long time ago.. Why this project is not promoted?

**Comment by RChrisCoble** [comment]
I'm not a contributor here or have a crystal ball, but I agree this project is obviously important enough to eventually be included in .Net, hopefully in the .Net 7 release.  .Net 6 is RC and now branched for .Net 7 development on main, so as this repo matures I hope it gets included alongside Maui.Graphics.  

As Maui.Graphics does not have a dependency on anything else "Maui", it's a fantastic way to write controls that can be used cross platform, and it's my understanding from watching a Xamarin community standup video there are Maui controls built on top of Maui.Graphics shipping in the .Net 6 timeframe.

**Reply by saint4eva** [reply]
> As Maui.Graphics does not have a dependency on anything else "Maui", it's a fantastic way to write controls that can be used cross platform, and it's my understanding from watching a Xamarin community standup video there are Maui controls built on top of Maui.Graphics shipping in the .Net 6 timeframe.

Maui.Graphics.Controls is built on top of Maui.Graphics. It would make it possible to write an app that looks the same on all platforms. And it would provide a performance boost since it is a drawn UI developed on top of a high performance graphichs library.

**Reply by sacdevops** [reply]
To release the Maui Controls with .NET 7 would be already to late. The community is giving the last chance for this project and team at 2Q22. When MAUI gets not better at this date as XF then the most people will switch to Flutter or other tools.
This is the last chance.of Microsoft.
And the Drawn Controls are the key for success. It has to be a part of .NET 6 and not 7, because it's a LTS. Many companies are not upgrading on Short Time Supports. When the Drawn Controls come with .NET 7, then it has not to be published anymore because the rest of the few thousands community will go also.

@jsuarezruiz know also that I'm one of the biggest critism for the .NET Maui Project and biggest Fan of the Drawn Controls + MVU. The only reason why my company and me not leaved already this community is the hope on these two features at 2Q22.

**Comment by jsuarezruiz** [comment]
Thanks for the feedback! At the moment it is only an experimental project where progress sometimes with the aim of doing tests and receiving feedback. However, it could be part for .NET 7 and have more support in the future.

**Comment by RChrisCoble** [comment]
Count me in as a request to productize this repo in .Net 7.  I'd like to start using it early next year to mix graphics and controls on the same canvas as we need that support directly with MAUI.Graphics.

**Comment by robloo** [comment]
Yes, this was the pretty wide consensus in https://github.com/dotnet/maui/issues/43. Devs have been asking for this forever. If Xamarin was designed like this from the start Flutter wouldn't be nearly as big as it is either. I kind-of doubt it will get any official traction until David Ortinau is replaced though -- he's blind to any feedback that doesn't follow his pre-existing viewpoints.

It's also unfortunate this is coming in so late considering the Uno Platform (which follows existing UWP XAML) and Avalonia which are doing wonderful work in this space already. A third entrant at this point will muddy things up a bit going forward. Still directionally this is finally the correct architecture from a Microsoft-led cross platform UI framework. 10 years late.

**Reply by sacdevops** [reply]
I share your opinion. David is the main reason Xamarin went under and .NET Maui will go under as well. I have been blocked by David, my comments have been deleted on Youtube just because I do not share his opinion and focus on the real problems of the project.
Just having Maui postponed to 2Q22 is a mess. Now should the Drawn Controls become part of .NET7? Do you want Maui to die out before it starts? The project grows on crap as long as you don't address the real problems.
I wish David would step down and someone more experienced and with vision would take the post.

**Reply by saint4eva** [reply]
> I share your opinion. David is the main reason Xamarin went under and .NET Maui will go under as well. I have been blocked by David, my comments have been deleted on Youtube just because I do not share his opinion and focus on the real problems of the project. Just having Maui postponed to 2Q22 is a mess. Now should the Drawn Controls become part of .NET7? Do you want Maui to die out before it starts? The project grows on crap as long as you don't address the real problems. I wish David would step down and someone more experienced and with vision would take the post.

Haven't your projects been postponed before? It seems you are complaining. If you feel other platforms are better for you, you could use them.

**Reply by Majestic-Mustang** [reply]
> Yes, this was the pretty wide consensus in [dotnet/maui#43](https://github.com/dotnet/maui/issues/43). Devs have been asking for this forever. If Xamarin was designed like this from the start Flutter wouldn't be nearly as big as it is either. I kind-of doubt it will get any official traction until David Ortinau is replaced though -- he's blind to any feedback that doesn't follow his pre-existing viewpoints.
> 
> It's also unfortunate this is coming in so late considering the Uno Platform (which follows existing UWP XAML) and Avalonia which are doing wonderful work in this space already. A third entrant at this point will muddy things up a bit going forward. Still directionally this is finally the correct architecture from a Microsoft-led cross platform UI framework. 10 years late.

@davidortinau You need to either listen to the developers or let someone with a vision replace you.

**Reply by tcsaddul** [reply]
Maybe Microsoft doesnt want to compete with UI vendors like Telerik or Syncfusion?

**Comment by najak3d** [comment]
We concur 100% with the sentiment, and view this library as what should be promoted as a "primary method for UI development" in MAUI.  Having a UI that looks identical on all OS platforms is ideal, for perhaps a majority of apps.   Plus it also enables integration into "games".

One of our primary objectives here is to enable MAUI to be rendered to a Bitmap rendered to an OpenGL surface as the HUD/UI.  Each MAUI layout (panel, stack, window) - would be associated to it's own Bitmap, which is then handed off to OpenGL to render to screen like a simple fast-updating Picture box (with transparency and click-through).   It's about time that Game Engines stop trying to re-invent their own custom UI's, and just allow everyone to use a common UI package, like MAUI.   MAUI rendering to a Unity3D app, is one of the home runs we hope to achieve here.

Note:  We've already accomplished this feat with Avalonia and Urho3D.NET.   But realize that the Home-Run solution isn't Avalonia -- but instead is MAUI.Graphics.Controls. 

Maui.Graphics.Controls -- is the key to this success, for Msft/MAUI.

We may be intending to help you complete the missing controls (e.g. List View, Tree View, Scroll View, ComboBox, ...).

Please - someone inside Msft, please shout at the Msft MAUI team, to see what it takes to put more priority onto this Graphics.Controls lib!   It's a VERY big deal!

===
NOTE: As quickly as we can do it, our intent is to produce a POC of this "rendering MAUI inside an OpenGL surface", to help boost the energy/priority of this concept.  We believe it should be a "front-page" feature of MAUI, with raised priority for development.   In conjunction with this, our current aim is to also make use of Comet.    

**Headline:   Maui.MVU rendering to Graphics.Controls Anywhere.  (even Games, and web pages)**

**Reply by ToolmakerSteve** [reply]
If I were Microsoft, I'd be spending as much money as it takes to make this a reality ASAP.

Microsoft's response to Flutter and ReactJS currently is Blazor. That's not enough. Without a much better answer, Microsoft will continue to lose developer mindshare. That answer is supposed to be MAUI - but I have concerns.

When MAUI becomes RC in 2022, if MAUI controls could run smoothly anywhere, including inside a browser, .NET would finally be a complete cross-platform environment.

I know some (many?) people will be happy with MAUI rendering to HTML controls - the Xamarin Forms approach of supporting native controls - but that isn't what I would like to use. I'd rather be completely free of platform-specific idiosyncrasies - peruse the [THREE THOUSAND OPEN Xamarin Forms ISSUES](https://github.com/xamarin/Xamarin.Forms/issues) to see the subtle problems when trying to map a cross-platform UI to native controls.

Please Microsoft, give us the OPTION to render to a bitmap, and avoid all those headaches.

tl;dr - this shouldn't be a mere "experimental" project. This should be one of the Top Priority projects that Microsoft is working on.

Coupled with a related Top Priority project, to make sure this works smoothly in a browser, under WASM. Then .NET can finally deliver on its full potential.

FYI, najak and I are going down this road - attempting to use this technology, first under desktop/mobile game engines, then in a browser. Would love to be in the loop re any progress in making this a first-class official part of MAUI.

**Reply by jsuarezruiz** [reply]
Again, and first of all, thanks for the feedback. @najak3d I appreciate your interest in helping this approach. Let me know if I can create issues and help to organize some tasks etc.

**Reply by najak3d** [reply]
> Again, and first of all, thanks for the feedback. @najak3d I appreciate your interest in helping this approach. Let me know if I can create issues and help to organize some tasks etc.

I spent many hours trying to get your current github master branch to compile/run.   I finally got it to compile at one point, but wouldn't run due to some missing platform.  So I responded to that by having Maui-check 0.10.0 do it's thing.   And after running many cycles of "dotnet install workload", and maui-check....  it still won't compile.  I'm on the latest VStudio 2022 Preview along with the latest .NET 6.0-Preview.  I've done many cycles of deleting bin/obj folders, and restarting Visual Studio.   Most of the time, when I restart Visual Studio, it just locks up and never shows the solution.  When VSTudio won't start, the fix is for me to delete obj/bin -then it'll start.

As of today, it compiles, but throws unhandled exception at startup.

"Unhandled exception at 0x00007FFF9984BE3E (KernelBase.dll) in GraphicsControls.Sample.WinUI.exe: 0xC000027B: An application-internal exception has occurred (parameters: 0x0000024202DD7140, 0x0000000000000002)."

===
Background:
We just found Maui.Graphics.Controls here a few days ago, and before that had concluded "we're going to need to create our own pixel-perfect solution for Maui ourselves" because of our conviction, that for many, this is a "silver bullet" for MAUI.   We personally, want our apps to work-the-same on all platforms.   We don't want three User Guides, or 3 user experiences, or the need to test/resolve issues that are platform-specific!   With Graphics.Controls solution-  once it works on Windows -- it works the same on iOS, Android, or Web! (without issue) 

My current App (with 7000 annual paying subscribers) has it's own custom UI system (in C#) that renders it's output to a single picture Box that fills the whole window.   From Xamarin Forms standpoint, our app consists of ONE PICTUREBOX -- and so our entire app simply blits itself to the picture box (double-buffered) for each frame, and it works flawlessly.    We dislike Xamarin.forms and XAML/MVVM, and are very much looking forward to migrating our development and apps to MAUI + MVU with Maui.Graphics.Controls for the UI.  

I've written UI systems several times before, and am willing to help bring this solution to completion.  (it needs ListView, TreeView, Dropdown, and Scrollable Views, for starters)    I've been doing C# almost solely since 2003.  I'm 51 yrs old now, with some extra time on my hands for open source.

===
What do we need?  

Please provide guidance on how to go about working with the master branch.    We'd like to be working alongside you, in the same environment.

**Reply by najak3d** [reply]
Maui-check 0.10.0 has some issues, here is the full output in txt file.   NOTE, I've run it 3 more times here -- always with same results and failures....  there is no change happening now when I run it.

I've done the two "dotnet install workload xxx" recommendations at the bottom, manually -- twice.

And still, I'm unable to get the master branch to run "Samples.UI" project.


[mgc-check-results.txt](https://github.com/dotnet/Microsoft.Maui.Graphics.Controls/files/7688131/mgc-check-results.txt)

**Reply by najak3d** [reply]
In Visual Studio 2022 Preview, at the top it reads warning "Visual Studio ran into an unexpected problem, and may affect some project system functions. Open log file "

Here is the log file.
[VsProjectFault_7700fdc2-8ddc-48fe-9d06-a6c88fbcbd66.failure.txt](https://github.com/dotnet/Microsoft.Maui.Graphics.Controls/files/7688139/VsProjectFault_7700fdc2-8ddc-48fe-9d06-a6c88fbcbd66.failure.txt)

**Reply by najak3d** [reply]
> Again, and first of all, thanks for the feedback. @najak3d I appreciate your interest in helping this approach. Let me know if I can create issues and help to organize some tasks etc.

Can you give me an ETA on when it's safe again to build this project?  (or a few hints in a "Getting Started" video)

ATM, none of our team can get this to build, using Maui-check 0.10.0 and manual installs.   Everything currently failing.

So we're likely to redirect ourselves for a time onto other tasks so that we don't waste all of our time trying to accomplish the impossible for something that should be relatively easy.   We assume that soon (when?), the coast will be clear for us to try this again, and then to have success.

We don't have to be working on latest right now -- we just need some recent version that is stable.   We'll do whatever you recommend as safest to "get started".   We'd like to proceed with our POC of integrating Maui into Urho.NET and Unity3D.  We've already got this working with Avalonia - but we still believe Maui is the future, and so want to put our eggs into the Maui basket, and help pave the wave for bringing more dev's back to Xamarin/Maui/.NET.

**Reply by robloo** [reply]
@najak3d 

> but we still believe Maui is the future

I'm curious why you believe Maui is better (will be better) than Avalonia. Could you elaborate on the benefits of Maui? From a technical standpoint Avalonia is still far more powerful and the supported controls are quite a bit better. The only issues I've seen are the development speed, documentation and support. I am very curious to know your issues with Avalonia to add to a comparison document I have.

**Reply by najak3d** [reply]
> @najak3d
> I'm curious why you believe Maui is better (will be better) than Avalonia. Could you elaborate on the benefits of Maui? From a technical standpoint Avalonia is still far more powerful and the supported controls are quite a bit better. The only issues I've seen are the development speed, documentation and support. I am very curious to know your issues with Avalonia to add to a comparison document I have.

Avalonia was a godsend, and is solid and able to be used.  But it's pure open-source with no one paid to make it work.

The biggest reason we're laying bets with MAUI is that this is a large Microsoft initiative, and Microsoft is investing into it.   And they seem to have "all the right objectives/goals", plus they have Comet for MVU and Graphics.Controls here for DrawnControls (like Avalonia).  

Additionally, Avalonia:
1. Currently is XAML/MVVM-only, as well, which is a detriment.     MAUI has Comet.
2. Source code seems to suffer similar slowness to Xamarin.Forms, due to it's reliance upon reflection/etc.  (e.g. changing Style on sample projects takes 3 seconds on a PC)
3. Source code is surprisingly complex.   Getting started with it has been quite bumpy so far.  For example, we're doing simple binding from their sample code, but in our project, the Command binding is failing.   And the code to debug it is quite deep/complex.

As for #3 - this may not be a true minus.   If we were convinced that MAUI was going to fail, then Avalonia becomes the natural runner-up here, for us.    I really don't want to be doing MVVM though;   MVU with C#-layouts is the way to go, IMO.

**Reply by robloo** [reply]
Thanks for summarizing this. I follow all the major cross-platform XAML solutions and this is good feedback to understand pros/cons.

> But it's pure open-source with no one paid to make it work.

The core Avalonia team formed a company and offer paid support. Several companies are using them now for production apps and it seems to be working just fine. My own company plans in some time to migrate over to Avalonia for desktop apps (away from WinUI). However, I'm keeping an eye on everything and MAUI is evolving into this space which I didn't expect a year ago.

> Source code seems to suffer similar slowness to Xamarin.Forms, due to it's reliance upon reflection/etc. (e.g. changing Style on sample projects takes 3 seconds on a PC)

I have never seen this on desktop platforms. Avalonia is the fastest cross-platform XAML framework I've seen and they've spent a lot of time optimizing hot code paths and removing both reflection and Linq to increase performance.

> Source code is surprisingly complex. 

I would say Avalonia source code is much less complex than both Xamarin Forms and Uno Platform -- it helps a lot when you don't have to do things for each platform separately. It does use a lot of reactive/MVVM concepts directly in the framework though I wish it didn't. There are a lot of undocumented differences from WPF although it behaves the same for the most part.

**Reply by najak3d** [reply]
We'll fallback to Avalonia if MAUI fails.   If MAUI succeeds, then it *should* become the preferred choice compared to Avalonia.  So we're just betting that Microsoft's paid-team of expert programmers can succeed, and beat out Avalonia.

Besides, we really want Comet/MVU.   There's an F# branch of Avalonia that is MVU, but it's not widely used and is F#. 

We'd like to write most of our UI's using C# markup style language.   Clancy seems to know what he's doing.

So, we're trying to HELP MAUI to succeed more and faster, because I think this is the optimum solution, compared to Avalonia.

**Reply by robloo** [reply]
Fair points and I can certainly understand where you're coming from. If Microsoft had this 5 years ago it wouldn't even be a question today and would have made a lot of decisions easier while saving a lot of time.

**Reply by najak3d** [reply]
Due to the utter failure we're experiencing over these past few weeks with the MAUI add-ons (Comet and Graphics.Controls), coupled with the lack of responsiveness of the MAUI team - has pushed us into the corner.  We are now moving ahead with our objectives using Avalonia + Urho.NET, to produce a Game Editor UI + Urho In-Game UI, both using Avalonia, simultaneously.   With Avalonia, with a few small (legit) hacks, it Just-Works.  And so with this easy success, are moving forward with Avalonia instead of MAUI.

Not sure if we'll be coming back to MAUI or not.   This was a big disappointment for us.

**Reply by najak3d** [reply]
For MAUI to succeed in doing what Avalonia is doing, this Graphics.Controls library is a *must*, so that we can render each MAUI window to a Bitmap, which is the rendered on a Urho Sprite (a UI element). 

And we are unable to make any progress with MAUI along these lines without the Graphics.Controls library, which for us, we've had zero luck getting it to compile/run from source.

I'm posting this here, to give anyone on the MAUI team a chance to prevent us from leaving MAUI in the next couple weeks.

**Reply by robloo** [reply]
Honestly, it might be better for Microsoft to embrace the existing Avalonia or even Uno Platform than create a 3rd entrant. Avalonia is by far the easiest to work with and most production ready. Its also fully rendered as you know.

That said, this here is very clearly an experimental repo with 1-2 contributors only. You expect too much. This was proof of concept and just testing. Please give them some slack, they have other jobs. This was never meant to even be trialed at a company at this point.

**Reply by najak3d** [reply]
We were willing to help finish the job here with Graphics.Controls, if we could get the original baseline to compile/run.

There's no "judgement" here against the team - so no slack to deliver.  We have timing needs for our project, and can't wait around to see if Maui actually makes these Graphic.Controls and Comet into something usable for production.   Since they were started a year ago, I was hoping for something more usable by now, so that we could start working with it ASAP.

So I'm just writing here to let them know that the current state of affairs is likely losing them an entire Gaming Platform that would have been pushing people back towards Maui, rather than Avalonia.  This is just honest feedback for the team to know this underlying cost to their delayed schedule.

**Reply by najak3d** [reply]
I will also admit that for the past couple weeks, being pushed back into the arms of Avalonia, has revealed more positive things about Avalonia, such as the the "Markup.Declarative" project that was just released, to make it far easier/cleaner to write succinct C# to build UI layouts.    Avalonia crowd seems filled with energetic, helpful, genius community members and contributors.  So they have really impressed us here too.

At this point, it seems that Avalonia is still in infancy regarding market share -- not a lot of production apps are using it.   It seems to be the best-of-its-kind, currently, and might end up becoming mainstream, with 1000's of apps using it, rather than just dozens.  (I don't know the official #, this is just a guess.)

**Reply by najak3d** [reply]
While for MAUI -- when they "get it right" (meaning COMET and Graphics.Controls are BOTH production) -- they'll surely drive 1000's of app to use it.   That's our assumption, and why we are willing to put our eggs into the MAUI basket if only they showed quicker/better progress at this time.   Things are instead, seeming somewhat abandoned for these vital add-on projects (DrawnControls + Comet).

**Reply by jsuarezruiz** [reply]
Updated the Wiki to include a detailed section about how to compile this project https://github.com/dotnet/Microsoft.Maui.Graphics.Controls/wiki/Build-the-project

**Reply by najak3d** [reply]
I will give it another try this week, and hopefully get it building/running.  Thank you! :)

**Reply by najak3d** [reply]
@jsuarezruiz ,
Do you have any plans to coordinate with Comet near-term to make sure both are using the same workload versions/etc?

Right now, I'm guessing they won't play together, because COMET's build instructions want us to use "preview" stuff, while yours just uses the non-preview.   (i.e. maui-check --preview, and .NET 6 Preview, etc -- is what Comet wants).

We really prefer  Comet-style C#-markup for UI creation, and so want to be using both.

But if we had to choose just one for now -- your DrawnControls, by far, is more important.  We can lay it all out with XAML for now.

**Reply by najak3d** [reply]
OK - on a NEW PC (opened box yesterday) - I followed all instructions in your build procedure, verbatim, and achieved "all looks good" from maui-check.  I have VS 2022 Preview 17.1.0 - Preview 1.1 installed.

When I created the sample Maui App, it won't build.   Produces 13 errors as shown here.  I realize this is more of a bug against the main MAUI team, but am posting it here, since it also causes this failure on your build instructions.

Next, I will re-attempt to build Graphics.Controls, and see what turns up.

![image](https://user-images.githubusercontent.com/10914515/147863934-03807928-4c56-467f-b899-a34fb24dedd4.png)

**Reply by najak3d** [reply]
So then I loaded up Graphics.Controls, per your instructions, and at first the GraphicsControls llibrary had an error and unloaded itself... not sure why.  But I selected "Reload project" and it reloaded.   Then I set Startup Project to WinUI, and F5 to compile/run.

It compiled fine, but when it ran, got this exception:
InvalidCastException - cast not valid.

![image](https://user-images.githubusercontent.com/10914515/147864040-e32ae17e-6661-4923-a7f0-6c4e0721d6af.png)

Stack trace shown here:

![image](https://user-images.githubusercontent.com/10914515/147864049-08a409aa-8230-455f-b0f6-aeca54fbcb74.png)

**Reply by najak3d** [reply]
First success - the GraphicsControlSample Console App deploys FINE to Android API 31 (Pixel 5) emulator.

So at least now we can view/edit these controls.  It's clear that the controls existing have some issues -- such as the sliders don't retain MouseDragFocus if the mouse goes outside of their bounds (this is not how these work everywhere else) -- so the fixes seem easy, and then need to add in the TreeView, dropdown, List view, etc.

We do want to be working on Windows, not Android for this, so we'd like to get the previous post's exception fixed (InvalidCastException).

**Reply by jsuarezruiz** [reply]
Updated dependencies, the sample project and Windows issues. Could you try again?

**Reply by najak3d** [reply]
It's on my agenda for tonight.

**Reply by najak3d** [reply]
After re-updating to the latest VS2022 Preview and .NET 6.101 (specifically was needed, and was not installed by VS2022 so got it manually).

Now it runs cleanly on Windows.   Very nice. 
Bugs:  The Editor and Entry boxes did not work -- couldn't focus/type anything.   All behaved disabled.
(I tried both the C# and XAML alternatives.)

I also noticed that the XAML makes use of styles:
    Style="{StaticResource InfoLabel}"

While for the C#, it just manually sets the properties.  It might be good to have the C# make use of something similar to Styles, rather than repeating the properties for FontSize and TextColor on every line.

Also - I think it would be very good to show a few types of data binding in this example code.

**Reply by jsuarezruiz** [reply]
Glad to read it works for you now on Windows. Will include C# styles and some data binding samples. Also, feel free to send PRs with changes in the samples, etc.

**Comment by timahrentlov** [comment]
I think the new Avalonia v11 delivers on the drawn controls promise. Especially with their "Avalonia MAUI Hybrid" sneak peak: https://youtu.be/qcZSr2ejH5I

This is pretty much what I've been waiting for.

**Reply by Pouria7** [reply]
I deployed the "Hello world" project of Avalonia (current stable version) on my Android phone yesterday.
I got the "Force close" Android error message.
Also the apk file size on single cpu architecture in release mode (default config) is about 30mb.

**Comment by coderbasket** [comment]
More avalonia here
https://coderbasket.github.io/framework/Avalonia

**Comment by coderbasket** [comment]
START UP SPEED COMPARISON 

Avalonia => 3.90 seconds (Debug empty)
Maui => 5.32 seconds  (Debug empty)
Flutter => 1.55 seconds (Release not empty)  Link [https://play.google.com/store/app](https://play.google.com/store/apps/details?id=com.achimsapps.templates)
